src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -506,169 +506,206 @@ public void CopyTo(Array array, int index)
                 {
-                    System.Threading.Interlocked.CompareExchange<Object>(ref _syncRoot, new object(), null);
+                    Threading.Interlocked.CompareExchange<object>(ref _syncRoot, new object(), null);
                 }
                 return _syncRoot;
             }
         }
 
-        public bool IsSynchronized
-        {
-            get
-            {
-                return false;
-            }
-        }
+        public bool IsSynchronized => false;
 
-        public bool IsReadOnly
-        {
-            get
-            {
-                return false;
-            }
-        }
+        public bool IsReadOnly => false;
 
-        public object Clone()
-        {
-            return new BitArray(this);
-        }
+        public object Clone() => new BitArray(this);
 
-        public IEnumerator GetEnumerator()
-        {
-            return new BitArrayEnumeratorSimple(this);
-        }
+        public IEnumerator GetEnumerator() => new BitArrayEnumeratorSimple(this);
 
         // XPerY=n means that n Xs can be stored in 1 Y. 
         private const int BitsPerInt32 = 32;
         private const int BytesPerInt32 = 4;
         private const int BitsPerByte = 8;
 
+        private const int BitShiftPerInt32 = 5;
+        private const int BitShiftPerByte = 3;
+        private const int BitShiftForBytesPerInt32 = 2;
+
         /// <summary>
         /// Used for conversion between different representations of bit array. 
-        /// Returns (n+(div-1))/div, rearranged to avoid arithmetic overflow. 
+        /// Returns (n + (32 - 1)) / 32, rearranged to avoid arithmetic overflow. 
         /// For example, in the bit to int case, the straightforward calc would 
-        /// be (n+31)/32, but that would cause overflow. So instead it's 
-        /// rearranged to ((n-1)/32) + 1, with special casing for 0.
+        /// be (n + 31) / 32, but that would cause overflow. So instead it's 
+        /// rearranged to ((n - 1) / 32) + 1.
+        /// Due to sign extension, we don't need to special case for n == 0, if we use
+        /// bitwise operations (since ((n - 1) >> 5) + 1 = 0).
+        /// This doesn't hold true for ((n - 1) / 32) + 1, which equals 1.
         /// 
         /// Usage:
-        /// GetArrayLength(77, BitsPerInt32): returns how many ints must be 
+        /// GetArrayLength(77): returns how many ints must be 
         /// allocated to store 77 bits.
         /// </summary>
         /// <param name="n"></param>
-        /// <param name="div">use a conversion constant, e.g. BytesPerInt32 to get
-        /// how many ints are required to store n bytes</param>
-        /// <returns></returns>
-        private static int GetArrayLength(int n, int div)
+        /// <returns>how many ints are required to store n bytes</returns>
+        private static int GetInt32ArrayLengthFromBitLength(int n)
+        {
+            Debug.Assert(n >= 0);
+            return ((n - 1) >> BitShiftPerInt32) + 1;
```

Comment:
This can be further optimized. The suggestion below is just two instructions: `lea` / `shr`.

```cs
return (int)((uint)(n - 1 + (1 << BitShiftPerInt32)) >> BitShiftPerInt32);
```

Reply:
> The suggestion below is just two instructions

[Sharplab](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKEwGZ8AmXAYVwG9zc3cAHAJwEsA3RHACm+AhFw8AdnFwBxIXACS0yjQCCXLogCeAGSGSA5nAAWAMS4B7VACEecfUdMB9S70NTEAGwAUUmZIAlKzsLGTsEfgA7Lg+PpK4ALS4BIG4AHzpuACsaQDUKaTh7AC+IWzlnLwCwqLi/nIKynCqGlp6BsbmVrb2jl3OkkIA7n7SuEGVYZHsmDFjcIFxAK7+SwnJBLgFPlsAPHs5gWmZR0URZWQlQA) is showing something different.

```text
C.GetInt32ArrayLengthFromBitLength_original(Int32)
    L0000: dec ecx
    L0002: sar ecx, 0x5
    L0005: lea eax, [rcx+0x1]
    L0008: ret

C.GetInt32ArrayLengthFromBitLength_new(Int32)
    L0000: add ecx, 0x1f
    L0003: mov eax, ecx
    L0005: shr eax, 0x5
    L0008: ret
```

Reply:
Sharplab is showing the disassembly for classic .NET only, I am not sure but the suggestion from @GrabYourPitchforks might work only for .NET Core.

Reply:
OK, BDN shows a difference:
https://www.diffchecker.com/1GPIRDWU

Will update.